Fix data loss bug

danielbeardsley opened the pull request:
Previously, hits received during a flush (after the /endpoint
request was sent, but before hearing the response) would be thrown
away because reset() was called when the response came back.

Now, we reset the data store right when we serialize it
(no data loss) and under normal circumstances, do nothing.
If there's an error during flush, we merge the old data store
with the new one.

I've also improved the test and changed flushing to consider
a non-200 response from the server as a failure.

danielbeardsley committed:
test: exit with 1 on failure

And lets extract a couple of functions while we're at it.

danielbeardsley committed:
test: simplify request sending

Why was 'zero' even there to begin with?

danielbeardsley committed:
test: assert long flush doesnt loose data

There's a bug in pixel-ping such that hits received between
the start of a flush and the response from the server are
thrown away. This test fails on that bug; a fix will come
in a further commit

danielbeardsley committed:
flush: fix data loss bug

Previously, hits received during a flush (after the /endpoint
request was sent, but before hearing the response) would be thrown
away because reset() was called when the response came back.

Now, we reset the data store right when we serialize it and
under normal circumstances, do nothing. If there's an error
and the config doesn't say we shouldn't, we merge the old
data store with the new one.

danielbeardsley committed:
flush: consider non-200 response as failed

Previously, a 400 or 500 level response to the flush endpint
would be considered valid.

Now, anything non-200 is considered a failure.

danielbeardsley commented:
CC @sctice @davidrans @anthonypileggi

djmetzle commented on src/pixel-ping.coffee:
```diff
@@ -29,7 +29,15 @@ serialize = ->
+# Merge the given `store` with the current one.
+merge = (new_store) ->
+  for key, hits of new_store
+    store[key] or= 0
```
Wat. Having a hard time with this. coffee seems really foreign.

danielbeardsley replied:
Yeah. coffee isn't as much fun as it was purported to be.

sctice replied:
I assume this is equivalent to `store[key] = store[key] || 0`. We use this logic twice now. It could be abstracted into something like `increment(n)`.

> GH wouldn't let me reply, so I'm editing your comment. Fixed!

djmetzle commented on src/pixel-ping.coffee:
```diff
@@ -45,15 +45,20 @@ merge = (new_store) ->
-  request = http.request endReqOpts, (response) ->
-    console.info '--- flushed ---'
+  request = http.request endReqOpts, (res) ->
+    if res.statusCode <= 299 and res.statusCode >= 200
```
Is everything in `200-299` valid?

I would expect you to explicitly want to check against ` == 200`?

https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml

danielbeardsley replied:
Everything in the 200 range is "Success". 200, 201, 204 are all very common success response codes (we produce them via our API). I don't see why we'd want to treat anything in the 200 range as failure.

djmetzle commented on src/pixel-ping.coffee:
```diff
@@ -45,15 +45,20 @@ merge = (new_store) ->
-    if !config.discard
-      merge(old_store)
-    console.log "--- cannot connect to endpoint : #{e.message}"
+    on_error "--- cannot connect to endpoint : #{e.message}"
```
Should `400` codes list `cannot connect`? Aren't they just generic application errors? Maybe would should `switch` out the response codes into classes?

danielbeardsley replied:
You're right, they shouldn't. I'm pretty sure the error event only happens on a network or protocol error.

From [the docs](https://nodejs.org/dist/latest-v6.x/docs/api/http.html):

> If any error is encountered during the request (be that with DNS resolution, TCP level errors, or actual HTTP parse errors) an 'error' event is emitted on the returned request object. As with all 'error' events, if no listeners are registered the error will be thrown.

djmetzle commented:
Woah. Coffeescript is gnarly. I had a hard time reading the commits for tests. I left some pesky comments. I think the logic here is pretty sound, differentiating and merging the count sets, and re-merging on failures. But i can't in good faith bless off on code i don't fully understand. 

Soft CR, i can dig deeper though if you need another :+1:

danielbeardsley commented:
It'd be great to get at least one more pair of eyes on this. Maybe @davidrans @sctice @sterlinghirsh

sctice commented on src/pixel-ping.coffee:
```diff
@@ -29,22 +29,36 @@ serialize = ->
   return unless config.endpoint
+  on_error = (message) ->
+    if !config.discard
+      merge(old_store)
```
It may not matter because of hoisting, but I'd really prefer to see `old_store = reset()` above the definition of `on_error`, where we reference `old_store`. The early reference to `old_store` makes me think that there's supposed to be an `old_store` global.

Also, I think `merge(old_store) unless config.discard` would read nicely.

danielbeardsley replied:
Fixed!

sctice commented on src/pixel-ping.coffee:
```diff
@@ -29,22 +29,36 @@ serialize = ->
-    reset()
-    console.info '--- flushed ---'
+  request = http.request endReqOpts, (res) ->
+    if res.statusCode <= 299 and res.statusCode >= 200
```
This could be nicely abstracted into an `isSuccessStatus` function (or something similar).

danielbeardsley replied:
Good point, fixed!

sctice commented on src/pixel-ping.coffee:
```diff
@@ -29,22 +29,36 @@ serialize = ->
 
 # Reset the `store`.
 reset = ->
+  old_store = store
```
The style in the rest of this package (and Node/JS in general) seems to be `camelCase`.

danielbeardsley replied:
Good point.

sctice commented on test/test-ping.coffee:
```diff
@@ -3,37 +3,65 @@ http        = require 'http'
       hits = JSON.parse data.json
-      if hits.one is 1 and hits.two is 2
-        console.log 'Test Succeeded'
+      if flush_assertions.shift() hits
```
The way `flush_and_assert` is written, it looks like pushing an assertion is optional, but the way this is written, it looks like we're expecting an assertion.

danielbeardsley replied:
Fixed!

sctice commented:
CR 🎏 but deploy_block ✋ on some suggestions. The only one I feel really strongly about is referencing `old_store` before it's defined in the local scope.

danielbeardsley committed:
s/old_store/oldStore/: most of the package uses camelCase

danielbeardsley committed:
oldStore: move store above reference

Functionally the same, it's just more palatable to see an assignment
before a reference, even if the ref is in a callback executed later.

danielbeardsley committed:
flush response: extract a success function

danielbeardsley committed:
tests: make callback non-optional

We always pass it anyway.

danielbeardsley committed:
Extract a function

danielbeardsley committed:
JS: update compiled JS file

After doing `cake build`

sctice commented:
This still adds a number of identifiers that are `snake_case` instead of `camelCase`, but I don't feel _that_ strongly about it. CR 🎏

danielbeardsley committed:
Variable names: use camelCase

Cause like thats like the standard in JS, bro.

danielbeardsley commented:
> This still adds a number of identifiers that are snake_case instead of camelCase

Fixed!

Derp, sorry I didn't fix that the first time.

sctice commented:
Thanks @danielbeardsley! CR 🎏

djmetzle commented:
That last commit is intense. CR :coffee:
